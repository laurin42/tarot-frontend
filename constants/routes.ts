import type { Route } from '@/types/navigation';

/**
 * Typsichere Routenkonstanten für die Navigation
 */
export const APP_ROUTES = {
  TABS: {
    THREECARDS: "/(tabs)/threecards" as Route,
    DAILYCARD: "/(tabs)/dailycard" as Route,
    PROFILE: "/(tabs)/profile" as Route,
  },
  AUTH: "/(auth)" as Route,
  HOME: "/" as Route,
};

/**
 * Konstanten für dynamische Routen
 */
export const DYNAMIC_ROUTES = {
  DEV_TOOLS: "/dev-tools" as const,
};