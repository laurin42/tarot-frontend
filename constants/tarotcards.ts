export interface ITarotCard {
  id: string;
  name: string;
  image: any;
}

export interface ISelectedAndShownCard extends ITarotCard {
  name: string;
  showFront?: boolean;
  isSelected?: boolean;
  currentIndex?: number;
  explanation?: string;
  image: any;
  onNextCard: () => void;
}

export const tarotCards: ITarotCard[] = [
  {
    id: "0",
    name:  "Der Narr",
    image: require("@/assets/images/tarot_cards/The_fool.jpg"),
  },
  {
    id: "1",
    name:  "Der Magier",
    image: require("@/assets/images/tarot_cards/The_magician.jpg"),
  },
  {
    id: "2",
    name: "Die Kaiserin",
    image: require("@/assets/images/tarot_cards/The_empress.jpg"),
  },
  {
    id: "3",
    name: "Die Hohepriesterin",
    image: require("@/assets/images/tarot_cards/High_priestess.jpg"),
  },
  {
    id: "4",
    name: "König der Kelche",
    image: require("@/assets/images/tarot_cards/King_of_cups.jpg"),
  },
  {
    id: "5",
    name: "Ritter der Kelche",
    image: require("@/assets/images/tarot_cards/Knight_of_cups.jpg"),
  },
  {
    id: "6",
    name: "Ritter der Münzen",
    image: require("@/assets/images/tarot_cards/Knight_of_pentacles.jpg"),
  },
  {
    id: "7",
    name: "Page der Kelche",
    image: require("@/assets/images/tarot_cards/Page_of_cups.jpg"),
  },
  {
    id: "8",
    name: "Page der Münzen",
    image: require("@/assets/images/tarot_cards/Page_of_pentacles.jpg"),
  },
  {
    id: "9",
    name: "Königin der Kelche",
    image: require("@/assets/images/tarot_cards/Queen_of_cups.jpg"),
  },
  {
    id: "10",
    name: "Zwei Kelche",
    image: require("@/assets/images/tarot_cards/Two_Cups.jpg"),
  },
  {
    id: "11",
    name: "Drei Kelche",
    image: require("@/assets/images/tarot_cards/Three_Cups.jpg"),
  },
  {
    id: "12",
    name: "Vier Kelche",
    image: require("@/assets/images/tarot_cards/Four_Cups.jpg"),
  },
  {
    id: "13",
    name: "Fünf Kelche",
    image: require("@/assets/images/tarot_cards/Five_Cups.jpg"),
  },
  {
    id: "14",
    name: "Sechs Kelche",
    image: require("@/assets/images/tarot_cards/Six_Cups.jpg"),
  },
  {
    id: "15",
    name: "Sieben Kelche",
    image: require("@/assets/images/tarot_cards/Seven_Cups.jpg"),
  },
  {
    id: "16",
    name: "Acht Kelche",
    image: require("@/assets/images/tarot_cards/Eight_Cups.jpg"),
  },
  {
    id: "17",
    name: "Neun Kelche",
    image: require("@/assets/images/tarot_cards/Nine_Cups.jpg"),
  },
  {
    id: "18",
    name: "Zehn Kelche",
    image: require("@/assets/images/tarot_cards/Ten_Cups.jpg"),
  },
  {
    id: "19",
    name: "Drei Münzen",
    image: require("@/assets/images/tarot_cards/Three_pents.jpg"),
  },
  {
    id: "20",
    name: "Vier Münzen",
    image: require("@/assets/images/tarot_cards/Four_pents.jpg"),
  },
  {
    id: "21",
    name: "Fünf Münzen",
    image: require("@/assets/images/tarot_cards/Five_pents.jpg"),
  },
  {
    id: "22",
    name: "Sechs Münzen",
    image: require("@/assets/images/tarot_cards/Six_pents.jpg"),
  },
  {
    id: "23",
    name: "Sieben Münzen",
    image: require("@/assets/images/tarot_cards/Seven_pents.jpg"),
  },
  {
    id: "24",
    name: "Acht Münzen",
    image: require("@/assets/images/tarot_cards/Eight_pents.jpg"),
  },
  {
    id: "25",
    name: "Neun Münzen",
    image: require("@/assets/images/tarot_cards/Nine_pents.jpg"),
  },
  {
    id: "26",
    name: "Zehn Münzen",
    image: require("@/assets/images/tarot_cards/Ten_pents.jpg"),
  },
  {
    id: "27",
    name: "Ass Münzen",
    image: require("@/assets/images/tarot_cards/Ace_of_Pentacles.jpg"),
  },
];


