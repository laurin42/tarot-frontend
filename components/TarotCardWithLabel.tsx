import React from "react";
import { View, Text, TouchableOpacity, Dimensions } from "react-native";
import OptimizedTarotCard from "@/components/OptimizedTarotCard";
import { Shadow } from "react-native-shadow-2";
import { ISelectedAndShownCard } from "@/constants/tarotcards";
import { commonStyles, globalTextStyles } from "@/styles/tarotTheme";
import { colors } from "@/styles/theme";

interface TarotCardWithLabelProps {
  card: ISelectedAndShownCard;
  index: number;
  onPress: (index: number) => void;
}

const TarotCardWithLabel: React.FC<TarotCardWithLabelProps> = ({
  card,
  index,
  onPress,
}) => {
  // Geräteabmessungen für responsive Darstellung
  const { width: screenWidth } = Dimensions.get("window");
  const cardWidth = screenWidth * 0.25; // ~25% mit etwas Abstand

  const getPositionLabel = (index: number): string => {
    switch (index) {
      case 0:
        return "Gegenwart";
      case 1:
        return "Konflikt";
      case 2:
        return "Perspektive";
      default:
        return `Position ${index + 1}`;
    }
  };

  return (
    <View style={commonStyles.cardWrapper}>
      {/* Position Label */}
      <Text style={globalTextStyles.cardLabel}>
        ({getPositionLabel(index)})
      </Text>

      {/* Tarot Card Image */}
      <View
        style={{
          width: "100%",
          alignItems: "center",
          justifyContent: "center",
          marginVertical: 4,
        }}
      >
        <TouchableOpacity onPress={() => onPress(index)}>
          <Shadow
            distance={8}
            startColor={colors.purpleGlow}
            endColor="rgba(139, 92, 246, 0.0)"
            offset={[0, 0]}
          >
            <OptimizedTarotCard
              cardId={card.id}
              imageSource={card.image}
              isShown={true}
              size="small" // Für Listenansichten ist "small" effizienter
              name={card.name}
              style={{
                width: cardWidth,
                height: cardWidth * 1.6,
                borderRadius: 8,
              }}
            />
          </Shadow>
        </TouchableOpacity>
      </View>

      {/* Card Name */}
      <View
        style={{
          marginTop: 8,
          width: "100%",
          minHeight: 36,
          justifyContent: "center",
        }}
      >
        <Text
          style={globalTextStyles.cardName}
          numberOfLines={2}
          ellipsizeMode="tail"
        >
          {card.name}
        </Text>
      </View>
    </View>
  );
};

export default React.memo(TarotCardWithLabel); // Optimierung durch Memoization
