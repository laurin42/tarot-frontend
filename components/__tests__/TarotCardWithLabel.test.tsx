import React from "react";
import { render, fireEvent, screen } from "@testing-library/react-native";
import TarotCardWithLabel from "../TarotCardWithLabel";
import { ISelectedAndShownCard } from "@/constants/tarotCards";

// Mock child components to isolate the TarotCardWithLabel logic
jest.mock("@/components/DynamicTarotCard", () => {
  // Need to require react-native inside the mock factory
  const MockView = require("react-native").View;
  // Return a function component that renders a View with a testID
  return (props: any) => (
    <MockView testID="dynamic-tarot-card-mock" {...props} />
  );
});
jest.mock("react-native-shadow-2", () => ({
  Shadow: jest.fn(({ children }) => children), // Simple mock that renders children
}));

// --- Mock Data ---
const mockCard: ISelectedAndShownCard = {
  id: "card-test",
  name: "The Magician",
  image: 2, // Example image number
  showFront: true,
  isSelected: false,
  onNextCard: jest.fn(),
};

const mockOnPress = jest.fn();
const testIndex = 1;

describe("TarotCardWithLabel Component", () => {
  beforeEach(() => {
    mockOnPress.mockClear();
    // Mock Dimensions.get only if not already handled globally by jest-expo
    // (Assuming jest-expo provides a basic mock)
    // jest.spyOn(Dimensions, 'get').mockReturnValue({ width: 375, height: 667, scale: 2, fontScale: 1 });
  });

  it("should render the card name correctly", () => {
    render(
      <TarotCardWithLabel
        card={mockCard}
        index={testIndex}
        onPress={mockOnPress}
      />
    );
    expect(screen.getByText(mockCard.name)).toBeTruthy();
  });

  // Removed test for the position label as it's no longer displayed
  // it('should render the correct position label', () => { ... });

  it("should render the DynamicTarotCard component", () => {
    render(
      <TarotCardWithLabel
        card={mockCard}
        index={testIndex}
        onPress={mockOnPress}
      />
    );
    // Check if the mocked component is present via testID
    expect(screen.getByTestId("dynamic-tarot-card-mock")).toBeTruthy();
  });

  it("should call onPress with the correct index when pressed", () => {
    render(
      <TarotCardWithLabel
        card={mockCard}
        index={testIndex}
        onPress={mockOnPress}
      />
    );

    // Find the mocked component via testID
    const cardElement = screen.getByTestId("dynamic-tarot-card-mock");
    // Fire press event on the element within the TouchableOpacity
    fireEvent.press(cardElement);

    expect(mockOnPress).toHaveBeenCalledTimes(1);
    expect(mockOnPress).toHaveBeenCalledWith(testIndex);
  });

  // Optional Snapshot Test
  it("should match snapshot", () => {
    const tree = render(
      <TarotCardWithLabel
        card={mockCard}
        index={testIndex}
        onPress={mockOnPress}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
