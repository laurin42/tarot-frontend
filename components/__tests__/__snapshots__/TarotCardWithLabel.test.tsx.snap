// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TarotCardWithLabel Component should match snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
      },
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "flexDirection": "column",
        "height": "auto",
        "justifyContent": "flex-start",
        "marginHorizontal": 2,
        "maxWidth": "30%",
        "width": "29%",
      },
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "marginVertical": 4,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        cardName="The Magician"
        isShown={true}
        style={
          {
            "borderRadius": 8,
            "height": 300,
            "width": 187.5,
          }
        }
        testID="dynamic-tarot-card-mock"
      />
    </View>
  </View>
  <View
    style={
      {
        "justifyContent": "center",
        "marginTop": 8,
        "minHeight": 36,
        "width": "100%",
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      style={
        {
          "color": "#A78BFA",
          "fontSize": 14,
          "fontWeight": "bold",
          "lineHeight": 18,
          "textAlign": "center",
        }
      }
    >
      The Magician
    </Text>
  </View>
</View>
`;
